lib/OpenQA/Worker/Job.pm

```diff
@@ -801,63 +821,70 @@ sub _upload_results_step_0_post_status {
+
+                $file = "$fileprefix/.thumbs/$file";
+                if (-f $file) {
+                    $self->_optimize_image($file);
+                    $form{file}->{file} = $file;
+                    $form{thumb} = 1;
+                    my $tx = $ua->post($ua_url => form => \%form);
+                    if (my $err = $tx->error) { log_error("Upload failed: $err") }
+                }
+            }
 
-    my $pooldir        = $self->worker->pool_directory;
-    my $fileprefix     = "$pooldir/testresults";
-    my $images_to_send = $self->images_to_send;
-    for my $md5 (keys %$images_to_send) {
-        my $file = $images_to_send->{$md5};
-        $self->_optimize_image("$fileprefix/$file");
-
-        log_debug("Uploading $file as $md5");
-        my %form = (
-            file => {
-                file     => "$fileprefix/$file",
-                filename => $file
-            },
-            image => 1,
-            thumb => 0,
-            md5   => $md5
-        );
-        # don't use api_call as it retries and does not allow form data
-        $tx = $ua->post($ua_url => form => \%form);
-        # FIXME: error handling?
-
-        $file = "$fileprefix/.thumbs/$file";
-        if (-f $file) {
-            $self->_optimize_image($file);
-            $form{file}->{file} = $file;
-            $form{thumb}        = 1;
-            $tx                 = $ua->post($ua_url => form => \%form);
-        }
-    }
-    $self->{_images_to_send} = {};
-
-    for my $file (@{$self->files_to_send}) {
-        log_debug("Uploading $file");
-        my %form = (
-            file => {
-                file     => "$pooldir/testresults/$file",
-                filename => $file,
-            },
-            image => 0,
-            thumb => 0,
-        );
-        # don't use api_call as it retries and does not allow form data
-        # (refactor at some point)
-        $tx = $ua->post($ua_url => form => \%form);
-        # FIXME: error handling?
-    }
-    $self->{_files_to_send} = [];
-    return !$tx || !$tx->error;
+            for my $file (@{$self->files_to_send}) {
+                log_debug("Uploading $file");
+                my %form = (
+                    file => {
+                        file     => "$pooldir/testresults/$file",
+                        filename => $file,
+                    },
+                    image => 0,
+                    thumb => 0,
+                );
+                # don't use api_call as it retries and does not allow form data
+                # (refactor at some point)
+                my $tx = $ua->post($ua_url => form => \%form);
+                if (my $err = $tx->error) { log_error("Upload failed: $err") }
+            }
+        },
+        sub {
+            my ($subprocess, $err) = @_;
+            log_error("Upload images subprocess error: $err") if $err;
+            $self->{_images_to_send} = {};
+            $self->{_files_to_send}  = [];
+            $callback->();
```

Comment:
So here the bare callback invocation is acceptable? (Instead of using `Mojo::IOLoop->next_tick($callback);`?)

Reply:
Yes, it's in the second `subprocess` callback, which happens after the subprocess is finished. Guaranteed to be at least one reactor tick further (and usually many more).

Reply:
If we have a clean event loop application a construct like `some_function(sub { $callback->() })` should always be safe.